fs: improve readFile performance

BridgeAR opened the pull request:
This increases the maximum buffer size per read to 256kb when using
`fs.readFile`. This is important to improve the read performance for
bigger files.

Refs: https://github.com/nodejs/node/issues/25741

Benchmark (512kb):

```
17:32:35                                                  confidence improvement accuracy (*)    (**)   (***)
17:32:35  fs/readfile.js concurrent=10 len=1024 dur=5                    -3.29 %       ±4.88%  ±6.48%  ±8.40%
17:32:35  fs/readfile.js concurrent=10 len=16777216 dur=5        ***    592.66 %      ±18.04% ±24.13% ±31.68%
17:32:35  fs/readfile.js concurrent=1 len=1024 dur=5                      2.18 %       ±4.41%  ±5.84%  ±7.57%
17:32:35  fs/readfile.js concurrent=1 len=16777216 dur=5         ***    356.82 %      ±25.96% ±34.72% ±45.57%
```

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

mscdex commented:
Couldn't we just change `kReadFileBufferLength` instead?

BridgeAR commented:
> Couldn't we just change kReadFileBufferLength instead?

That would not have the expected behavior: it is now only used in case the `stat` call failed and we do not know the file size. In that case, it's likely best to be conservative and not to pre-allocate a big buffer.

mscdex commented:
Maybe we should have two constants then, since `kReadFileBufferLength` would no longer be completely accurate?

antsmartian commented on lib/internal/fs/read_file_context.js:
```diff
@@ -79,7 +79,8 @@ class ReadFileContext {
       offset = this.pos;
-      length = Math.min(kReadFileBufferLength, this.size - this.pos);
+      // Read up to 256kb.
+      length = Math.min(256 * 1024, this.size - this.pos);
```
If I understand this correctly, this should work fine, but may be we need to add a way for the user to define the min buffer size (if no options, fall back to `256kb`)? Will that make sense?

BridgeAR replied:
I would rather not do this in this PR. I personally just want to make the default better. The only reason for a user to set this value manually would be for big files. If a user wants this extra performance boost they should be aware of the trade off with the thread pool and I doubt that this is the case for most developers. If someone indeed knows about this trade off, they can also load the file with other `fs` functions directly.

sam-github commented:
> That would not have the expected behavior: it is now only used in case the stat call failed and we do not know the file size. In that case, it's likely best to be conservative and not to pre-allocate a big buffer.

It would have the behaviour I'd expect from the commit description! :-)

Why is this more conservative? What is safer? If we are going to read 1,000K, and we know it, with this change it will be read in larger 256K chunks. But if we don't know that it will be 1,000K, it will be read in the older, smaller, 8K chunks. Maybe there is a good reason for this, but its not immediately obvious.

I'm actually not clear on why there are two branches at all. It seems like the branch where size isn't known would work for all cases. AFAICT the only optimization is that if the file happens to be exactly a multiple of kReadFileBufferLength, and .size is known, one extra `read()` (which would just return `0` meaning EOF), is avoided. Is that the entire point? But if .size came from a `stat()` call, then we always pay the `stat()` price, and almost never get the "1 less read" bonus, so it doesn't seem a net win.

.... OK, I'm not deleting the above, but I think I just realized why there are two branches. The `.size !== 0` branch uses `pread()` in libuv, which doesn't change the file position, but that won't work if a fd was passed, and the fd isn't seekable (its a stream or pipe, like stdin). Its trying to make `fs.readFile(A_FD)` not change the current file position. I think a comment to this effect, if my guess is right, would be useful.

Still not sure why the larger buf size isn't used in all cases.

BridgeAR commented:
@mscdex 
> Maybe we should have two constants then, since kReadFileBufferLength would no longer be completely accurate?

I added another commit that also refactors the code a tiny bit to make it simpler. I also added the constant as suggested.

@sam-github 
> I think a comment to this effect, if my guess is right, would be useful.

There is a comment in one branch that checks for `size === 0`. Should I add another one for this specific case?

mscdex commented on lib/internal/fs/read_file_context.js:
```diff
@@ -73,9 +72,10 @@ class ReadFileContext {
     if (this.size === 0) {
-      buffer = this.buffer = Buffer.allocUnsafeSlow(kReadFileBufferLength);
+      buffer = Buffer.allocUnsafeSlow(kReadFileUnknownBufferLength);
+      this.buffer = buffer;
```
minor nit: maybe this could be moved to after `length = ` to more closely match the order of assignments in the `else`?

jasnell commented:
There's a slightly different approach we could take on this by either:

a. Providing an option to set the read chunk size... e.g. `fs.readFile(fn, { readSize: 10000 }, () = > {})`, or...

b. Providing an option that is effectively a preallocated buffer to fill on each read... `fs.readFile(fn, { readBuf: Buffer.from(10000) }, () => {})` 

Option a is likely the far better choice and allows users to performance tune on their own.

BridgeAR commented:
I increased the value to 512kb and also doubled the value for unknown file sizes.

256kb already show significant improvements:

```
15:41:32                                                  confidence improvement accuracy (*)    (**)   (***)
15:41:32  fs/readfile.js concurrent=10 len=1024 dur=5                     0.61 %       ±4.89%  ±6.48%  ±8.38%
15:41:32  fs/readfile.js concurrent=10 len=16777216 dur=5        ***    504.18 %      ±10.37% ±13.83% ±18.07%
15:41:32  fs/readfile.js concurrent=1 len=1024 dur=5                     -0.86 %       ±4.08%  ±5.40%  ±6.99%
15:41:32  fs/readfile.js concurrent=1 len=16777216 dur=5         ***    285.81 %      ±17.78% ±23.72% ±31.01%
```

@jasnell I believe we should raise this limit with or without the option. So adding the option could be done in a separate PR and this is a quick win.

zbjornson commented:
https://github.com/nodejs/node/compare/master...zbjornson:chunksizeopt this branch has a `chunkSize` option like @jasnell's (a) if it's useful. I was using it for benchmarking.

However, @davisjam and I actually just had a nice chat and agreed that partitioning may be causing more harm than good (in part due to the increased risk of OOMs), and were both open to removing partitioning and instead adding a prominent doc warning.

BridgeAR commented:
@jasnell 
> b. Providing an option that is effectively a preallocated buffer to fill on each read... fs.readFile(fn, { readBuf: Buffer.from(10000) }, () => {})

So that the buffer would be as big as the whole file? We would have to guard against wrong sizes (and then potentially still manage multiple buffers) and what would be the benefit of b over a?

jasnell commented:
As I said, option a is likely better than option b ;-)

sam-github commented:
> There is a comment in one branch that checks for size === 0. 

If you mean

> Unknown size, just read until we don't get bytes.

That says what, but not why. Since you can read until you don't get bytes whether you know the size or not, without some context, its quite mysterious why the file doesn't just always use the `size===0` branch. I happen to remember the kerflufle about using fds instead of file names, so I think I guessed why, but only after writing a paragraph about how I didn't undertstand. It shouldn't take insider knowledge to understand this.

This was not introduced in this PR, so consider it a nit. I can fix, if you want.

BridgeAR commented:
@sam-github 

This is not about using an fd or not but about the file type.

We know the file size for regular files (S_IFREG type) but not in case it's another file type. So if we know the file type, it'll allocate the full buffer in advance and then read in chunks into that buffer. For other file types we allocate a small buffer because we could otherwise over allocate and that costs memory and performance. 

See https://github.com/nodejs/node/pull/27063/files#diff-9a205ef7ee967ee32efee02e58b3482dR265

BridgeAR commented:
@nodejs/fs this could use some further reviews.

@sam-github @jasnell @zbjornson I suggest to land this, no matter what the latter solution might be to improve the current situation. Are you good with that?

BridgeAR commented:
@mcollina 
> LGTM. However I’d make both of them 512KB.

I guess it won't make a huge difference as the cases where the file does not have a regular file type is not common but it will definitely take a significantly longer time to allocate 512kb over the current 8kb and if we over allocate, we also do work and use more memory than we should. What about using 64kb or 128kb for unknown file sizes? I guess with that size we cover a lot more files and also improve the performance significantly.

mcollina commented:
> I guess it won't make a huge difference as the cases where the file does not have a regular file type is not common but it will definitely take a significantly longer time to allocate 512kb over the current 8kb and if we over allocate, we also do work and use more memory than we should. What about using 64kb or 128kb for unknown file sizes? I guess with that size we cover a lot more files and also improve the performance significantly.

Mine is not a blocker. I'm ok as it is. Maybe can you add a comment with the above?

BridgeAR commented:
I added a comment to highlight the difference and updated the buffer size to 64kb for unknown file sizes.

BridgeAR commented:
Landed in eb2d4161f56efa7d82a6364734c674757f1f3ea4 :tada:

Thanks for the reviews!
